models/user.go

```diff
@@ -520,5 +520,87 @@ func (u *User) IsBanned() bool {
 
 func (u *User) UpdateBannedUntil(tx *storage.Connection) error {
 	return tx.UpdateOnly(u, "banned_until")
+}
+
+// RemoveUnconfirmedIdentities removes potentially malicious unconfirmed identities from a user (if any)
+func (u *User) RemoveUnconfirmedIdentities(tx *storage.Connection) error {
+	unconfirmedProviders := []string{
+		"email",
+		"phone",
+	}
+
+	providersToRemove := make(map[string]bool)
+
+	for _, unconfirmedProvider := range unconfirmedProviders {
+		switch unconfirmedProvider {
+		case "email":
+			if u.IsConfirmed() {
+				continue
+			}
+
+			u.EncryptedPassword = ""
+
+			if terr := tx.UpdateOnly(u, "encrypted_password"); terr != nil {
+				return terr
+			}
+
+		case "phone":
+			if u.IsPhoneConfirmed() {
+				continue
+			}
+
+			u.Phone = ""
+
+			if terr := tx.UpdateOnly(u, "phone"); terr != nil {
+				return terr
+			}
+
+		default:
+			continue
+		}
+
+		providersToRemove[unconfirmedProvider] = true
```

Comment:
nit: the nested `continue` statements inside the `switch` block makes this slightly hard to read because initially i thought that this line would always run

Reply:
Yeah it is pretty terrible, I'll try to find another way.